File: mobile/ios/ShiftPlanner/ShiftPlanner/Features/Employee/Availability/Repositories/APIAvailabilityRepository.swift
import Foundation

final class APIAvailabilityRepository: AvailabilityRepository {
    private let apiClient: APIClient

    init(apiClient: APIClient = .shared) {
        self.apiClient = apiClient
    }

    func fetchAvailability(employeeId: Int) async throws -> EmployeeAvailabilityResponseDTO {
        let request = apiClient.makeRequest(
            path: "employees/\(employeeId)/availability",
            method: "GET",
            requiresAuthorization: true
        )
        return try await apiClient.send(request, as: EmployeeAvailabilityResponseDTO.self)
    }

    func saveAvailability(employeeId: Int, payload: EmployeeAvailabilityUpsertDTO) async throws -> EmployeeAvailabilityResponseDTO {
        let body = try JSONEncoder().encode(payload)
        let request = apiClient.makeRequest(
            path: "employees/\(employeeId)/availability",
            method: "POST",
            body: body,
            requiresAuthorization: true
        )
        return try await apiClient.send(request, as: EmployeeAvailabilityResponseDTO.self)
    }
}

File: mobile/android/app/src/main/java/com/froggyriia/shiftplanner/data/availability/ApiAvailabilityRepository.kt
package com.froggyriia.shiftplanner.data.availability

import com.froggyriia.shiftplanner.data.network.ApiClient
import com.froggyriia.shiftplanner.data.network.EmployeeAvailabilityBlockDto
import com.froggyriia.shiftplanner.data.network.EmployeeAvailabilityResponseDto
import com.froggyriia.shiftplanner.data.network.EmployeeAvailabilityUpsertDto
import com.froggyriia.shiftplanner.domain.model.AvailabilityBlock
import com.froggyriia.shiftplanner.domain.model.AvailabilitySlotState
import com.froggyriia.shiftplanner.domain.model.EmployeeAvailabilityData

class ApiAvailabilityRepository(
    private val apiClient: ApiClient
) : AvailabilityRepository {

    override suspend fun fetchAvailability(employeeId: Int): EmployeeAvailabilityData = wrap {
        apiClient.api.getEmployeeAvailability(employeeId).toDomain()
    }

    override suspend fun saveAvailability(
        employeeId: Int,
        data: EmployeeAvailabilityData
    ): EmployeeAvailabilityData = wrap {
        apiClient.api.saveEmployeeAvailability(
            id = employeeId,
            request = EmployeeAvailabilityUpsertDto(
                weeklyAvailability = data.weeklyAvailability.map { block ->
                    EmployeeAvailabilityBlockDto(
                        weekday = block.weekday,
                        startTime = block.startTime,
                        endTime = block.endTime,
                        availabilityStatus = block.status.apiValue
                    )
                },
                desiredDaysOff = data.desiredDaysOff
            )
        ).toDomain()
    }

    private fun EmployeeAvailabilityResponseDto.toDomain(): EmployeeAvailabilityData =
        EmployeeAvailabilityData(
            employeeId = employeeId,
            weeklyAvailability = weeklyAvailability.map { block ->
                AvailabilityBlock(
                    weekday = block.weekday,
                    startTime = block.startTime,
                    endTime = block.endTime,
                    status = AvailabilitySlotState.fromApiValue(block.availabilityStatus)
                )
            },
            desiredDaysOff = desiredDaysOff
        )

    private suspend fun <T> wrap(block: suspend () -> T): T {
        return try {
            block()
        } catch (error: Throwable) {
            throw Exception(apiClient.userMessage(error))
        }
    }
}
